// @statsfm/spotify.js/dist/lib/SpotifyAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpotifyAPI = void 0;
const AlbumManager_1 = require("./album/AlbumManager");
const ArtistManager_1 = require("./artist/ArtistManager");
const AudioManager_1 = require("./audio/AudioManager");
const MeManager_1 = require("./me/MeManager");
const PlaylistManager_1 = require("./playlist/PlaylistManager");
const SearchManager_1 = require("./search/SearchManager");
const TrackManager_1 = require("./track/TrackManager");
const UserManager_1 = require("./user/UserManager");
class SpotifyAPI {
    constructor(config) {
        this.privateConfig = {};
        this.config = config;
        this.tracks = new TrackManager_1.TrackManager(this.config, this.privateConfig);
        this.albums = new AlbumManager_1.AlbumManager(this.config, this.privateConfig);
        this.artists = new ArtistManager_1.ArtistManager(this.config, this.privateConfig);
        this.users = new UserManager_1.UserManager(this.config, this.privateConfig);
        this.me = new MeManager_1.MeManager(this.config, this.privateConfig);
        this.search = new SearchManager_1.SearchManager(this.config, this.privateConfig);
        this.audio = new AudioManager_1.AudioManager(this.config, this.privateConfig);
        this.playlist = new PlaylistManager_1.PlaylistManager(this.config, this.privateConfig);
    }
}
exports.SpotifyAPI = SpotifyAPI;
