// @statsfm/spotify.js/dist/interfaces/Errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BadRequestError = exports.InternalServerError = exports.RatelimitError = exports.NotFoundError = exports.ForbiddenError = exports.AuthError = void 0;
class AuthError extends Error {
    constructor(message, stack) {
        super(message);
        this.name = 'AuthError';
        if (stack)
            this.stack = stack;
    }
}
exports.AuthError = AuthError;
class ForbiddenError extends Error {
    constructor(message, stack) {
        super(message);
        this.name = 'ForbiddenError';
        if (stack)
            this.stack = stack;
    }
}
exports.ForbiddenError = ForbiddenError;
class NotFoundError extends Error {
    constructor(message, stack) {
        super(message);
        this.name = 'NotFoundError';
        if (stack)
            this.stack = stack;
    }
}
exports.NotFoundError = NotFoundError;
class RatelimitError extends Error {
    constructor(message, stack) {
        super(message);
        this.name = 'RatelimitError';
        if (stack)
            this.stack = stack;
    }
}
exports.RatelimitError = RatelimitError;
class InternalServerError extends Error {
    constructor(message, stack) {
        super(message);
        this.name = 'InternalServerError';
        if (stack)
            this.stack = stack;
    }
}
exports.InternalServerError = InternalServerError;
class BadRequestError extends Error {
    constructor(message, stack) {
        super(message);
        this.name = 'BadRequestError';
        if (stack)
            this.stack = stack;
    }
}
exports.BadRequestError = BadRequestError;
